turbinia/lib/libcloudforensics.py

```diff
@@ -554,6 +614,109 @@ def get_source_string(self):
+    Returns:
+      an API operation object for the referenced compute resource
+    Raises RuntimeError:
+      if resource type is not defined as a type which extends the
+      GoogleGomputeBaseResource class
+    """
+    resource_type = self.get_resource_type()
+    module = None
+    if resource_type == 'compute#instance':
+      module = self.project.gce_api().instances()
+    elif resource_type == 'compute#disk':
+      module = self.project.gce_api().disks()
+    elif resource_type == 'compute#snapshot':
+      module = self.project.gce_api().snapshots()
+    else:
+      raise RuntimeError((
+          'Compute resource Type {0:s} is not on of the defined types in '
+          'libcloudforensics library (Instance, Disk or Snapshot) '
+      ).format(resource_type))
+
+    operation_func_to_call = getattr(module, operation_name)
+    return operation_func_to_call
+
+  def get_labels(self):
+    """Get all labels of a compute resource.
+
+    Returns:
+      A dictionary of all labels.
+    """
+
+    operation = self.get_operation().execute()
+
+    resource_response = self.project.gce_operation(
+        operation, zone=self.zone, block=False)
+    return resource_response.get('labels')
+
+  def add_labels(self, new_labels_dict):
+    """Add or update labels of a compute resource.
+
+    Args:
+      new_labels_dict: dictionary containing the labels to be added. ex:
+        {"incident_id":"1234abcd"}
+
+    Returns:
+      response of the API operation
+    """
+
+    get_operation = self.get_operation().execute()
+    labelFingerprint = self.project.gce_operation(
+        get_operation, zone=self.zone, block=False)['labelFingerprint']
+
+    exisitng_labels_dict = dict()
+    if self.get_labels() is not None:
+      exisitng_labels_dict = self.get_labels()
+    exisitng_labels_dict.update(new_labels_dict)
+    labels_dict = exisitng_labels_dict
+    request_body = {'labels': labels_dict, 'labelFingerprint': labelFingerprint}
+
+    resource_type = self.get_resource_type()
+    operation = None
+    if resource_type == 'compute#instance':
+      operation = self.form_operation('setLabels')(
+          instance=self.name, project=self.project.project_id, zone=self.zone,
+          body=request_body).execute()
+    elif resource_type == 'compute#disk':
+      operation = self.form_operation('setLabels')(
+          resource=self.name, project=self.project.project_id, zone=self.zone,
+          body=request_body).execute()
+    elif resource_type == 'compute#snapshot':
+      operation = self.form_operation('setLabels')(
+          resource=self.name, project=self.project.project_id,
+          body=request_body).execute()
+    else:
+      raise RuntimeError((
+          'Compute resource Type {0:s} is not on of the defined types in '
+          'libcloudforensics library (Instance, Disk or Snapshot) '
+      ).format(resource_type))
+
+    return self.project.gce_operation(operation, zone=self.zone, block=False)
```

Comment:
It might be better to leave the choice to the caller whether this will be blocking? It can be tricky not to block by default if the caller uses label information right away.

Reply:
yes will add it as an optional argument to the add_labels method with default equal False and pass it to the request